### PR TITLE
enable tablets in kespaces created by node.stress

### PR DESCRIPTION
**Pull request by soyacz**
We want to test tablets functionality without need for test adaptation (creating keyspace with tablets enabled  before node.stress is run).

Modify `stress_options` by add `initial_tablets` option and change replication strategy to `NetworkTopologyStrategy` to support tablets tests.

fixes: https://github.com/scylladb/scylla-ccm/issues/509

### CODE REVIEW

File: ccmlib/node.py [hunk 1]
```diff
@@ -197,6 +198,9 @@ def is_scylla(self):
     def is_docker():
         return False
 
+    def tablets_enabled(self) -> bool:
+        return 'tablets' in self.cluster._config_options.get('experimental_features', [])
```
Comment (bhalevy) [inline ccmlib/node.py:202]: nit: this function should be implemented in cluster.py and be called here

File: ccmlib/node.py [hunk 2]
```diff
@@ -1277,7 +1281,10 @@ def stress_process(self, stress_options, **kwargs):
                 # specify used jmx port if not already set
                 if not [opt for opt in stress_options if opt.startswith('jmx=')]:
                     stress_options.extend(['-port', 'jmx=' + self.jmx_port])
-
+        if self.tablets_enabled():
+            stress_options = self._set_keyspace_initial_tablets(
+                stress_options, datacenter=self.data_center, rf=min(len(self.cluster.nodes), 3)
```
Comment (bhalevy) [inline ccmlib/node.py:1286]: maybe be the default rf should be calculated this way in `_set_keyspace_initial_tablets`
so the logic could be reused in the future if and when called from other places

File: ccmlib/node.py [hunk 3]
```diff
@@ -1295,6 +1302,41 @@ def stress(self, stress_options=None, capture_output=False, **kwargs):
...
+        default_replication = f"replication(strategy=NetworkTopologyStrategy,{datacenter}={rf},initial_tablets={initial_tablets})"
+        try:
+            idx = stress_options.index('-schema')
+            replication = stress_options[idx + 1]
+            if 'replication' in replication:
+                if 'NetworkTopologyStrategy' in replication:
+                    # if NTP already set, just prepend initial_tablets
+                    replication = re.sub(r'replication\(.+?\)',
+                                         lambda m: m.group(0).replace(')', f',initial_tablets={initial_tablets})'), replication)
+                else:
```
Comment (bhalevy) [inline ccmlib/node.py:1327]: shouldn't we check for `SimpleStrategy`?
what if it's `EverywhereStrategy`?
Maybe the funtion should throw an error in this case

Comment (soyacz) [reply ccmlib/node.py:1327]: currently anything but NetworkStrategy will be overridden completely. I don't know much about EverywhereStrategy - is it available and can be used by user keyspaces?

File: ccmlib/node.py (hunk 2 above)
Comment (soyacz) [reply ccmlib/node.py]: yes, at first I wanted also this way, but to ease unit testing i made it this way. I moved this calculation to `cluster` object. WDYT?

File: ccmlib/node.py (hunk 3 above)
Comment (fruch) [reply ccmlib/node.py:1325]: and more, can you pass initial_tablets on `EverywhereStrategy` and `SimpleStrategy` ?

File: setup.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
 
 setup(
     name='ccm',
-    version='2.0.5',
+    version='2.0.6',
```
Comment (fruch) [inline setup.py:19]: why changing this ? i.e. we are not uploading CCM anywhere, no point of changing this

File: tests/test_set_keyspace_initial_tablets.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import pytest
+from typing import List, Optional
+
+from ccmlib.node import Node
+
+
+@pytest.mark.parametrize("stress_options, initial_tablets, datacenter, rf, expected", [
+    # typical case
+    (['write', 'n=0', 'no-warmup', '-schema', 'replication(factor=1)', '-rate', 'threads=1'], 8, None, 3,
+     ['write', 'n=0', 'no-warmup', '-schema', 'replication(strategy=NetworkTopologyStrategy,datacenter1=1,initial_tablets=8)', '-rate', 'threads=1']),
```
Comment (fruch) [inline tests/test_set_keyspace_initial_tablets.py:10]: looking at those example, I don't think we should be changing the `strategy`, that might be changing the test logic more then we intend.

SimpleStrategy should accept `initial_tablets`, and if not, I think it's a bug

File: ccmlib/node.py (hunk 3 above)
Comment (soyacz) [reply ccmlib/node.py:1325]: `SimpleStrategy` fails with `ConfigurationException: Unrecognized strategy option {initial_tablets} passed to org.apache.cassandra.locator.SimpleStrategy for keyspace keyspace1` 
`EverywhereStrategy` - I can create but it ignores tablets option (it is mentioned in describe keyspace, but `initial_tablets` don't change anything in sstables structure).

File: setup.py (hunk 4 above)
Comment (soyacz) [reply setup.py:19]: is this causing any problems? 
I mean, when I reinstall ccm I always see 2.0.5 and i wonder if there was some change...

File: tests/test_set_keyspace_initial_tablets.py (hunk 5 above)
Comment (soyacz) [reply tests/test_set_keyspace_initial_tablets.py:10]: `initial_tablets` are not planned for `SimpleStrategy`, at least not in near future. 

I don't see any other option than change this to `NetworkTopologyStrategy`.
Shall we split this change to first adapt dtests to NTS?

File: setup.py (hunk 4 above)
Comment (fruch) [reply setup.py:19]: It's irrelevant to this change

Since we don't have releases.

We can open an issue for starting doing so, I would love putting CCM up in pypi, it can help us in lots of places.

But for now we stick to branches.

Comment (soyacz) [reply setup.py]: fixed.
I also consider puting ccm to pypi as a good idea.

File: ccmlib/cluster.py [hunk 6]
```diff
@@ -720,3 +720,11 @@ def error(self, message):
     @staticmethod
     def is_docker():
         return False
+
+    @property
+    def tablets_enabled(self) -> bool:
+        return 'tablets' in self._config_options.get('experimental_features', [])
+
+    @property
```
Comment (fruch) [inline ccmlib/cluster.py:728]: I think we should do the c-s default, which is 1

File: ccmlib/node.py (hunk 3 above)
Comment (fruch) [reply ccmlib/node.py:1325]: EverywhereStrategy is used in 3 dtest modules only, non of them uses c-s, we should check if c-s even supports using it

File: ccmlib/cluster.py (hunk 6 above)
Comment (soyacz) [reply ccmlib/cluster.py]: true. Fixed.

**Review by fruch — APPROVED** [review]
LGTM

File: ccmlib/node.py [hunk 7]
```diff
@@ -1295,6 +1300,41 @@ def stress(self, stress_options=None, capture_output=False, **kwargs):
         except KeyboardInterrupt:
             pass
 
+    @staticmethod
+    def _set_keyspace_initial_tablets(stress_options: List[str], initial_tablets:int = 8,
```
[the remark this answers is not in the document]
Comment (fruch) [reply ccmlib/node.py:1304]: Is there some reference to how one should pick this number ?

[the remark this answers is not in the document]
Comment (avikivity) [reply ccmlib/node.py:1304]: Minimum: 1*(total number of vcpus in the cluster) (suitable for up to 100 tables)
Maximum: 33*(total number of vcpus in the cluster) (suitable for a small number of tables)

For ccm you can treat the boundaries as fairly fluid. Better to have too many than too few. Eventually it will just be a hint and the load balancer will recover from bad inputs.

[the remark this answers is not in the document]
Comment (soyacz) [reply ccmlib/node.py:1304]: ccm usecase is running scylla on local machines, usually they are not too powerful (and run mutliple scylla nodes) and clusters are not loaded too much with data.
Also, this is just 'initial' number of tablets and in future split/merge will be implemented and this setting will not have that much impact I believe.

I can leave as is currently or implement some logic behind like:
`self._smp * 8` and then to match the nearest bigger power of 2 (IIRC tablets number must be always power of 2):
```
import math

def multiply_and_find_nearest_power_of_two(smp):
     smp *= 8
     smp = math.pow(2, math.ceil(math.log(smp, 2)))
     return smp

```

[the remark this answers is not in the document]
Comment (fruch) [reply ccmlib/node.py:1304]: @soyacz where this 8 is coming from ? @avikivity suggested thet the maximum should be used.

[the remark this answers is not in the document]
Comment (avikivity) [reply ccmlib/node.py:1304]: Note you shouldn't use smp directly, rather the sum of all smp values across the cluster (smp * nodes).

And it doesn't need to be exact.

[the remark this answers is not in the document]
Comment (soyacz) [reply ccmlib/node.py:1308]: I updated the code with higher number (smp * 32, coerced up to power of 2)

**Review by fruch — APPROVED** [review]
LGTM

**Comment by fruch** [comment]
@bhalevy @avikivity, care to give it a final review ?

File: ccmlib/node.py [hunk 8]
```diff
@@ -1277,7 +1279,14 @@ def stress_process(self, stress_options, **kwargs):
                 # specify used jmx port if not already set
                 if not [opt for opt in stress_options if opt.startswith('jmx=')]:
                     stress_options.extend(['-port', 'jmx=' + self.jmx_port])
-
+        if self.cluster.tablets_enabled:
+            smp = sum([node._smp or 1 for node in self.cluster.nodes.values()])
+            # default initial tablets: 32 * `sum of SMPs in cluster` coerced to power of 2 (suitable for small number of tables)
+            initial_tablets = int(math.pow(2, math.ceil(math.log(smp*32, 2))))
+            self.debug("tablets enabed, adjusting stress options by replication strategy change.")
```
Comment (avikivity) [inline ccmlib/node.py:1286]: enabled

File: ccmlib/node.py [hunk 9]
```diff
@@ -1277,7 +1279,14 @@ def stress_process(self, stress_options, **kwargs):
...
+            stress_options = self._set_keyspace_initial_tablets(
+                stress_options, datacenter=self.data_center, rf=1, initial_tablets=initial_tablets
```
Comment (avikivity) [inline ccmlib/node.py:1288]: Why rf=1? seems unrelated here.

Comment (fruch) [reply ccmlib/node.py:1288]: It's just used to setup a default which matches cassandra stress default which if you don't define the schema option it would be SimpleStrategy with rf=1

I agree it's a bit weird to have it in the arguments if it's a default.

Comment (soyacz) [reply ccmlib/node.py]: updated

File: ccmlib/node.py (hunk 8 above)
Comment (soyacz) [reply ccmlib/node.py]: updated
